File: OOP/Week 3/3.2 Clock/Clock/Clock/Clock.cs
using System;
namespace ClockNamspace
{
    public class Clock
    {
        private Counter seconds;
        private Counter minutes;
        private Counter hours;

        public int Seconds { get => seconds.Ticks; }
        public int Minutes { get => minutes.Ticks; }
        public int Hours   { get => hours.Ticks; }

        public Clock()
        {
            seconds = new Counter("Seconds");
            minutes = new Counter("Minutes");
            hours   = new Counter("Hours");
        }

        public Clock(string time)
            : this()
        {
            // split string into hours, minutes and seconds
            string[] segements = time.Split(':');

            // check that there aren't too many segments
            // 'Split()' always returns at least 1 string
            if (segements.Length > 3)
            {
                throw new ArgumentException("Invalid time format");
            }

            switch (segements.Length)
            {
                case 3:
                    if (int.TryParse(segements[0], out int h))
                    {
                        hours = new Counter("Hours", h);
                        goto case 2;
                    }
                    break;
                case 2:
                    if (int.TryParse(segements[segements.Length - 2], out int m))
                    {
                        minutes = new Counter("Minutes", m);
                        goto case 1;
                    }
                    break;
                case 1:
                    if (int.TryParse(segements[segements.Length - 1], out int s))
                    {
                        seconds = new Counter("Seconds", s);
                        return;
                    }
                    break;
            }

            throw new ArgumentException("Invalid time format");
        }

        public void Tick()
        {
            seconds.Increment();

            if (seconds.Ticks >= 60)
            {
                seconds.Reset();
                minutes.Increment();
            }

            if (minutes.Ticks >= 60)
            {
                minutes.Reset();
                hours.Increment();
            }

            if (hours.Ticks >= 24)
            {
                hours.Reset();
            }
        }

        public void Reset()
        {
            seconds.Reset();
            minutes.Reset();
            hours.Reset();
        }

        public override string ToString()
        {
            string s = Seconds.ToString("00");
            string m = Minutes.ToString("00");
            string h = Hours.ToString("00");
            return string.Format("{0}:{1}:{2}", h, m, s);
        }
    }
}

File: OOP/Week 10/10.1 - Clock in Another Language/clock.rb
class Counter
  attr_accessor :name
  def initialize(name, ticks = 0)
    @count = ticks
    @name = name
  end

  def ticks
    @count
  end

  def increment
    @count += 1
  end

  def reset
    @count = 0
  end
end

class Clock
  def initialize(time = nil)
    @secs = Counter.new("secs")
    @mins = Counter.new("mins")
    @hours = Counter.new("hours")

    if time != nil
      set_from_time(time)
      return
    end
  end

  def set_from_time(time)
    segments = time.split(':')
    raise "Invalid time format" if segments.size > 3
    if segments.size == 3
      @hours = Counter.new("hours", segments[0].to_i)
    end
    if segments.size >= 2
      @mins = Counter.new("mins", segments[-2].to_i)
    end
    if segments.size >= 1
      @secs = Counter.new("secs", segments[-1].to_i)
    end
  end

  def tick
    @secs.increment
    if @secs.ticks >= 60
      @secs.reset
      @mins.increment
    end
    if @mins.ticks >= 60
      @mins.reset
      @hours.increment
    end
    if @hours.ticks >= 24
      @hours.reset
    end
  end

  def reset
    @secs.reset
    @mins.reset
    @hours.reset
  end

  def to_s
    "%02d:%02d:%02d" % [@hours.ticks, @mins.ticks, @secs.ticks]
  end
end

def main
  clock = Clock.new("23:59:59")
  clock.tick
  puts clock
end

if __FILE__ == $0
  main
end
